core/src/main/scala/kafka/server/KafkaConfig.scala

```diff
@@ -1653,13 +1673,27 @@ class KafkaConfig(val props: java.util.Map[_, _], doLog: Boolean, dynamicConfigO
     }
   }
 
+  def controllerListenerNames: Seq[String] = {
+    val controllerListeners = getList(KafkaConfig.ControllerListenersProp)
+    if (controllerListeners == null) {
+      Seq.empty
+    } else {
+      controllerListeners.asScala.toSeq
+    }
+  }
+
   def dataPlaneListeners: Seq[EndPoint] = {
-    Option(getString(KafkaConfig.ControlPlaneListenerNameProp)) match {
-      case Some(controlPlaneListenerName) => listeners.filterNot(_.listenerName.value() == controlPlaneListenerName)
-      case None => listeners
+    listeners.filterNot { listener =>
+      val name = listener.listenerName.value()
+      name.equals(getString(KafkaConfig.ControlPlaneListenerNameProp)) ||
+        controllerListenerNames.contains(name)
```

Comment:
Will this cause the `controllerListenerNames` to be triggered multiple times?

Reply:
It will, but it's not a big concern since this only happens during initialization, and N is really small (like 3 or 4)